src/filter.ts

```diff
@@ -0,0 +1,159 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {Operator, Filter as IFilter} from './query';
+import {entity} from './entity';
+
+enum CompositeOperator {
+  AND = 'AND',
+  OR = 'OR',
+}
+
+export function AND(filters: EntityFilter[]): CompositeFilter {
+  return new CompositeFilter(filters, CompositeOperator.AND);
+}
+
+export function OR(filters: EntityFilter[]): CompositeFilter {
+  return new CompositeFilter(filters, CompositeOperator.OR);
+}
+
+/**
+ * A Filter is a class that contains data for a filter that can be translated
+ * into a proto when needed.
+ *
+ * @see {@link https://cloud.google.com/datastore/docs/concepts/queries#filters| Filters Reference}
+ *
+ */
+export abstract class EntityFilter {
+  /**
+   * Gets the proto for the filter.
+   *
+   */
+  // eslint-disable-next-line
+  abstract toProto(): any;
+}
+
+/**
+ * A PropertyFilter is a filter that gets applied to a query directly.
+ *
+ * @see {@link https://cloud.google.com/datastore/docs/concepts/queries#property_filters| Property filters Reference}
+ *
+ * @class
+ */
+export class PropertyFilter extends EntityFilter implements IFilter {
+  name: string;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  val: any;
+  op: Operator;
+
+  /**
+   * Build a Property Filter object.
+   *
+   * @param {string} Property
+   * @param {Operator} operator
+   * @param {any} val
+   */
+  constructor(property: string, operator: Operator, val: any) {
+    super();
+    this.name = property;
+    this.op = operator;
+    this.val = val;
+  }
+
+  private encodedValue(): any {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    let value: any = {};
+    if (this.name === '__key__') {
+      value.keyValue = entity.keyToKeyProto(this.val);
+    } else {
+      value = entity.encodeValue(this.val, this.name);
+    }
+    return value;
+  }
+
+  /**
+   * Gets the proto for the filter.
+   *
+   */
+  // eslint-disable-next-line
+  toProto(): any {
+    const OP_TO_OPERATOR = new Map([
```

Comment:
nit: I would probably hoist this to the top of the file, rather than declaring in `toProto`, at scale this will speed things up since you're allocating less `Maps`.